text_handler.py

```diff
@@ -26,42 +29,58 @@ async def __handleYoutubeDownload(self, message: types.Message):
         if not str(message.chat.id) in config.WHITELIST_CHAT_ID:
             return
         try:
-            yt = YouTube(message.text)
-            author = yt.author
-            length = yt.length
-            title = yt.title
-            stream = (yt.streams
-                      .filter(only_audio=True, file_extension='mp4')
-                      .order_by('bitrate').desc().first())
-            filesize = stream.filesize
-            video_id = extract.video_id(message.text)
-            video_url = "youtu.be/" + video_id
+            ydl_opts = {
+                    'writethumbnail': True,
+                    'outtmpl': 'downloads/%(extractor)s_%(id)s.%(ext)s',
+                    'format': 'bestaudio',
+            }
+            with YoutubeDL(ydl_opts) as ydl:
+                info = ydl.extract_info(message.text, download=False)
+                audio_url = format(info['url'])
             telegram_audio_limit = 52428800
+            request_head = Request(audio_url, method='HEAD')
+            audio_filesize = urlopen(request_head).headers['Content-Length']
 
-            if (filesize < telegram_audio_limit and length < 600):
-                output_audiofile = stream.download(output_path="downloads/",
-                                                   filename=title +
-                                                   " - YouTube - " +
-                                                   video_id)
-                output_thumbnail = "downloads/square_thumbnail_" + video_id + \
-                                   ".jpg"
-                self.__make_squarethumb(yt.thumbnail_url, output_thumbnail)
+            if int(audio_filesize) < telegram_audio_limit:
+                ydl.download([message.text])
```

Comment:
why not reuse the variable **audio_url** here?

Reply:
`audio-url` is the direct link for audio file. message.txt is link similar to webpage_url which is the youtube link.

with `ydl.download([message.txt])`, it downloads both the thumbnail and audio file. I can get `webpage_url` at first and then `ydl.download` but I think it looks better this way and the variable is not used too much.

btw, youtube-dl is being called twice, `ydl.extract_info` and `ydl.download`, ideally we can download it by ourself. but I didn't try that yet also idk if that's more complicated or could be simpler. you may try to improve this or leave it as is for now.

Reply:
calling twice puts us in a greater risk of getting `429 Too Many Requests`, right?

Reply:
In rare cases, when the thumbnail url does not work, no thumbnail will be downloaded, but it will still try to upload the audio and the thumbnail (which does not exist) and produce error. I've experienced it with [this soundcloud url](https://soundcloud.com/blackspacemusic/hollow-coves-the-woods-black-space-remix)

Reply:
> calling youtube-dl twice puts us in a greater risk of getting 429 Too Many Requests, right?

probably. I'll try to solve this. but uncommon requests may also get some issues, youtube-dl deal with user-agent stuff properly.

Reply:
Use urllib to download audio and thumbnail now.